aten/src/ATen/native/cpu/ConvolutionMM3d.cpp

```diff
@@ -0,0 +1,335 @@
+      groups,
+      true);
+
+  Tensor grad_weight_2d = view_weight_2d(grad_weight);
+  checkContiguous(c, grad_weight_arg);
+
+  auto grad_output_contiguous = grad_output.contiguous();
+
+  const int64_t batch_size = input.size(0);
+  Tensor finput = compute_columns3d(input, stride, padding, kernel_size, groups);
+
+  AT_DISPATCH_FLOATING_TYPES_AND(
+      kBFloat16, input.scalar_type(), "slow_conv3d_cpu_grad_weight", [&] {
+    auto grad_weight_2d_a = grad_weight_2d.accessor<scalar_t, 2>();
+    auto grad_output_a = grad_output_contiguous.accessor<scalar_t, 5>();
+    auto finput_a = finput.accessor<scalar_t, 3>();
+    for (const auto t : c10::irange(batch_size)) {
+      auto grad_output_t = grad_output_a[t];
+      auto finput_t = finput_a[t];
+      slow_conv3d_backward_weight_frame(
+          grad_weight_2d_a, grad_output_t, finput_t, groups);
+    }
+  });
+}
+
+std::tuple<Tensor&, Tensor&, Tensor&> slow_conv3d_backward_out_cpu(const Tensor& grad_output,
+    const Tensor& self,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    Tensor& grad_input,
+    Tensor& grad_weight,
+    Tensor& grad_bias) {
+  // TODO: hacky way of determine the group size
+  int64_t groups = self.size(1) / weight.size(1);
+  if (grad_input.defined()) {
+    slow_conv3d_backward_out_cpu_template(
+        grad_input,
+        grad_output,
+        self,
+        weight,
+        kernel_size,
+        stride,
+        padding,
+        groups);
+  }
+
+  if (grad_bias.defined()) {
+    at::sum_out(grad_bias, grad_output, IntArrayRef{0, 2, 3, 4});
+  }
+
+  if (grad_weight.defined()) {
+    grad_weight.resize_(weight.sizes());
+    grad_weight.zero_();
+    slow_conv3d_backward_parameters_out_cpu_template(
+        grad_weight,
+        self,
+        grad_output,
+        kernel_size,
+        stride,
+        padding,
+        groups);
+  }
+
+  return std::tuple<Tensor&, Tensor&, Tensor&>(
+      grad_input, grad_weight, grad_bias);
+}
+
+std::tuple<Tensor, Tensor, Tensor> slow_conv3d_backward_cpu(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    std::array<bool, 3> output_mask) {
+  Tensor grad_input;
+  Tensor grad_weight;
+  Tensor grad_bias;
+
+  if (output_mask[0]) {
+    grad_input = at::empty({0}, grad_output.options());
+  }
+
+  if (output_mask[1]) {
+    grad_weight = at::empty({0}, grad_output.options());
+  }
+
+  if (output_mask[2]) {
+    grad_bias = at::empty({0}, grad_output.options());
+  }
+
+  at::native::slow_conv3d_backward_out_cpu(
+      grad_output,
+      self,
+      weight,
+      kernel_size,
+      stride,
+      padding,
+      grad_input,
+      grad_weight,
+      grad_bias);
+
+  return std::make_tuple(grad_input, grad_weight, grad_bias);
+}
+
+} // namespace
+
+REGISTER_DISPATCH(slow_conv3d_backward_stub, &slow_conv3d_backward_cpu);
```

Comment:
I thought about it some more, and couldn't we do something like the following?
- keep slow_conv3d_backward_cpu in aten/native (and not in aten/native/cpu)
- Add a slow_conv3d_backward_stub that only has a CUDA implementation attached to it (and maybe a dummy CPU implementation if that's necessary)
- When we need to call "slow_conv3d_backward" from aten/native, first we check if the inputs are on CPU. If they are on CPU, then we invoke slow_conv3d_backward_cpu.
- If the inputs are not on CPU, then we invoke "slow_conv3d_backward_stub", which will call out to the CUDA version

This avoids any problems we have with compiling a file multiple times in native/cpu. Also, the consequences of unnecessarily putting something in native/cpu is more than just increased build times -- it is also larger binary sizes

Reply:
Cool good call, and it's even easier for the specific case of slow 3d because there's no CUDA kernel for it :)

Reply:
@zou3519 I've got a slightly different idea for avoiding recompilations: what if we introduce a new macro `REGISTER_ALL_CPU_DISPATCH` that registers the same kernel across all CPU arch types? We'd call this from `native/Convolution*.cpp` and wouldn't need to move any logic underneath the `native/cpu` dir. That would simplify these PRs quite a bit and would also avoid the recompilation. Wdyt about this approach?

Reply:
Just to check... for this specific case we don't need a dispatch macro, right? Because there is no CUDA implementation? If so you might want to change the PR body

Reply:
Yep that's right! will change